pkg/registry-disk/registry-disk_test.go

```diff
@@ -0,0 +1,196 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2017 Red Hat, Inc.
+ *
+ */
+
+package registrydisk
+
+import (
+	"io/ioutil"
+	"os"
+	"os/user"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	v1 "kubevirt.io/kubevirt/pkg/api/v1"
+	diskutils "kubevirt.io/kubevirt/pkg/ephemeral-disk-utils"
+)
+
+var _ = Describe("RegistryDisk", func() {
+	tmpDir, _ := ioutil.TempDir("", "registrydisktest")
+	owner, err := user.Current()
+	if err != nil {
+		panic(err)
+	}
+
+	VerifyDiskType := func(diskExtension string) {
+		vm := v1.NewMinimalVM("fake-vm")
+		vm.Spec.Domain.Devices.Disks = append(vm.Spec.Domain.Devices.Disks, v1.Disk{
+			Type:   "RegistryDisk:v1alpha",
+			Device: "disk",
+			Source: v1.DiskSource{
+				Name: "someimage:v1.2.3.4",
+			},
+			Target: v1.DiskTarget{
+				Device: "vda",
+			},
+		})
+
+		// create a fake disk file
+		volumeMountDir := generateVMBaseDir(vm)
+		err = os.MkdirAll(volumeMountDir+"/disk0", 0750)
+		Expect(err).ToNot(HaveOccurred())
+
+		filePath := volumeMountDir + "/disk0/disk-image." + diskExtension
+		_, err := os.Create(filePath)
+
+		vm, err = MapRegistryDisks(vm)
+		Expect(err).ToNot(HaveOccurred())
+
+		// verify file gets renamed by virt-handler to prevent container from
```

Comment:
I don't fully understand that comment.

Reply:
Ha, you hit a deep topic. I'll explain what's happening and why. 

When the RegistryDisk container copies the disk file to a shared directory, that container is responsible for removing the file before it is allowed to exit.

So, the logic is
- copy disk
- go into wait loop periodically checking disk copy exists
- on exit signal, remove disk if it exists before exiting
- when disk is no longer present, exit. 

Now while all that is going on in the RegistryDisk container, when virt-handler gains knowledge of the presence of a local RegistryDisk, virt-handler takes ownership of the disk from the container by renaming the file.  At that point, virt-handler owns the lifecycle of the disk.  The RegistryDisk container will detect the disk it was watching "disappeared" because of the rename and exit.  If the ownership change never occurs for whatever reason, the RegistryDisk container will still own the disk and remove the disk when it is asked to terminate.

Here's why we do this.

For graceful shutdown, it's important that virt-handler takes ownership of the disk from the RegistryDisk container before the VM starts so the disk's lifecycle is not tied to the RegistryDisk container's lifecycle. Otherwise we have to synchronize guaranteeing the RegistryDisk's container will not terminate before the VM process or the virt-launcher container.  This is easily solved by having virt-launcher own the disk before starting the VM. 

Essentially, having virt-handler take ownership of the disk ensures it has complete control over the disk's lifecycle without having to compete (or attempt to synchronize) with the timing of the RegistryDisk containers termination.